satellite/repair/checker/observer.go

```diff
@@ -0,0 +1,396 @@
+// Copyright (C) 2022 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package checker
+
+import (
+	"context"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/spacemonkeygo/monkit/v3"
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+
+	"storj.io/common/storj"
+	"storj.io/common/uuid"
+	"storj.io/storj/satellite/metabase/rangedloop"
+	"storj.io/storj/satellite/metabase/segmentloop"
+	"storj.io/storj/satellite/overlay"
+	"storj.io/storj/satellite/repair"
+	"storj.io/storj/satellite/repair/queue"
+)
+
+var _ rangedloop.Observer = (*RangedLoopObserver)(nil)
+
+// RangedLoopObserver implements the ranged loop Observer interface.  Should be renamed to checkerObserver after rangedloop will replace segmentloop.
+//
+// architecture: Observer
+type RangedLoopObserver struct {
+	logger               *zap.Logger
+	repairQueue          queue.RepairQueue
+	nodestate            *ReliabilityCache
+	statsCollector       *statsCollector
+	repairOverrides      RepairOverridesMap
+	nodeFailureRate      float64
+	repairQueueBatchSize int
+	startTime            time.Time
+	TotalStats           aggregateStats
+
+	// Remote segments over threshold counters, to avoid panics during parallel Processing.
+	counter1 *monkit.Counter
+	counter2 *monkit.Counter
+	counter3 *monkit.Counter
+	counter4 *monkit.Counter
+	counter5 *monkit.Counter
+}
+
+// NewRangedLoopObserver creates new checker observer instance.
+func NewRangedLoopObserver(logger *zap.Logger, repairQueue queue.RepairQueue, overlay *overlay.Service, config Config) rangedloop.Observer {
+	return &RangedLoopObserver{
+		logger: logger,
+
+		repairQueue:          repairQueue,
+		nodestate:            NewReliabilityCache(overlay, config.ReliabilityCacheStaleness),
+		statsCollector:       newStatsCollector(),
+		repairOverrides:      config.RepairOverrides.GetMap(),
+		nodeFailureRate:      config.NodeFailureRate,
+		repairQueueBatchSize: config.RepairQueueInsertBatchSize,
+
+		TotalStats: aggregateStats{},
```

Comment:
nit: this initialization could be elided since it will already be the zero value